### PR TITLE
VN barcamp : Enable optional JWT authentication on WebAdmin endpoints

### PR DESCRIPTION
**Pull request by chibenwa**
We want the web application to be able to access James Admin API from the internet.

To do so, we need it to be secure.

We will implement an optionnal mechanism for authentication based on JWT. We will check the Bearer header, with configured public key, and ensure it contains the claims admin=true.

### CODE REVIEW

File: server/protocols/webadmin/src/test/java/org/apache/james/webadmin/authentication/JwtFilterTest.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+                return haltException.statusCode() == 401;
+            }
+            return false;
+        }
+
+        @Override
+        public void describeMismatch(Object o, Description description) {}
+
+        @Override
+        public void _dont_implement_Matcher___instead_extend_BaseMatcher_() {}
```
Comment (mbaechler) [inline server/protocols/webadmin/src/test/java/org/apache/james/webadmin/authentication/JwtFilterTest.java]: (:

File: server/container/guice/jpa-guice/src/main/java/org/apache/james/JPAJamesServerMain.java [hunk 2]
```diff
@@ -45,7 +46,8 @@ public static void main(String[] args) throws Exception {
         GuiceJamesServerImpl server = new GuiceJamesServerImpl()
                     .combineWith(jpaServerModule, 
                             new JMXServerModule(), 
-                            new LuceneSearchMailboxModule());
+                            new LuceneSearchMailboxModule())
+                    .overrideWith(new NoWebAdminFilterModule());
```
Comment (mbaechler) [inline server/container/guice/jpa-guice/src/main/java/org/apache/james/JPAJamesServerMain.java]: why it needs to be disabled on JPA ?

File: server/protocols/webadmin/src/test/java/org/apache/james/webadmin/authentication/JwtFilterTest.java (hunk 1 above)
Comment (chibenwa) [inline server/protocols/webadmin/src/test/java/org/apache/james/webadmin/authentication/JwtFilterTest.java]: Ahh

File: server/container/guice/jpa-guice/src/main/java/org/apache/james/JPAJamesServerMain.java (hunk 2 above)
Comment (chibenwa) [inline server/container/guice/jpa-guice/src/main/java/org/apache/james/JPAJamesServerMain.java]: No JMAP on top of JPA => no JMAP

I don't want to have dilemna of "should we create a separate config file for jwt" so I decided to not some it during barcamp and keep for later.

Comment (mbaechler) [inline server/container/guice/jpa-guice/src/main/java/org/apache/james/JPAJamesServerMain.java]: I understand the jwt config stuff, but what is the link with JMAP, we were talking about WebAdmin, right ? WebAdmin has to be available for most backend, don't you think ? (I agree is out of barcamp scope, obviously)

Comment (chibenwa) [inline server/container/guice/jpa-guice/src/main/java/org/apache/james/JPAJamesServerMain.java]: It's for now enableable on JPA but not with JWT...

File: server/protocols/jwt/src/main/java/org/apache/james/jwt/JwtTokenVerifier.java [hunk 3]
```diff
@@ -40,11 +44,16 @@
     }
 
     public boolean verify(String token) throws JwtException {
-        String subject = extractLogin(token);
-        if (Strings.isNullOrEmpty(subject)) {
-            throw new MalformedJwtException("'subject' field in token is mandatory");
+        try {
```
Comment (quynhn) [inline server/protocols/jwt/src/main/java/org/apache/james/jwt/JwtTokenVerifier.java]: We catch JwtException so I think we don't need to define to throws it. WDYT?

File: server/protocols/jwt/src/main/java/org/apache/james/jwt/JwtTokenVerifier.java [hunk 4]
```diff
@@ -54,12 +63,18 @@ public String extractLogin(String token) throws JwtException {
                 .getSubject();
     }
 
-    public void hasAttribute(String attributeName, Object expectedValue, String token) throws JwtException {
-        Jwts
-            .parser()
-            .require(attributeName, expectedValue)
-            .setSigningKey(pubKeyProvider.get())
-            .parseClaimsJws(token);
+    public boolean hasAttribute(String attributeName, Object expectedValue, String token) throws JwtException {
```
Comment (quynhn) [inline server/protocols/jwt/src/main/java/org/apache/james/jwt/JwtTokenVerifier.java]: idem

File: server/protocols/jwt/src/test/java/org/apache/james/jwt/JwtTokenVerifierTest.java [hunk 5]
```diff
@@ -141,9 +134,9 @@ public void extractLoginShouldWorkWithAdminClaim() {
 
     @Test
     public void hasAttributeShouldThrowIfClaimInvalid() throws Exception {
-        expectedException.expect(ClaimJwtException.class);
+        boolean authorized = sut.hasAttribute("admin", true, VALID_TOKEN_ADMIN_FALSE);
```
Comment (quynhn) [inline server/protocols/jwt/src/test/java/org/apache/james/jwt/JwtTokenVerifierTest.java]: Naming function? ShouldThrow? but it does not

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/JwtFilterIntegrationTest.java [hunk 6]
```diff
@@ -117,4 +116,17 @@ public void jwtShouldRejectNonValidRequests() throws Exception {
             .doesNotContain(DOMAIN);
     }
 
+    @Test
+    public void jwtShouldRejectNonValidRequests() throws Exception {
```
Comment (quynhn) [inline server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/JwtFilterIntegrationTest.java]: s/NonValid/Invalid?

**Comment by chibenwa** [comment]
Merged in linagora:vn_barcamp

**Pull request closed without merging**